raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java

```diff
@@ -0,0 +1,178 @@
+            // For other exception during state file loading, we could assume the file is corrupted already
+            // and should be cleaned up.
+            delete();
+            log.info("Local quorum file corrupted, deleting {}", file.toPath());
+            return recoverStateFromLog(file, logEndOffsetAndEpoch);
+        }
+    }
+
+    private QuorumStateData recoverStateFromLog(File file, final OffsetAndEpoch logEndOffsetAndEpoch) throws IOException {
+        QuorumStateData stateFromMetadataLog = new QuorumStateData();
+        incorporateElectionState(stateFromMetadataLog,
+            ElectionState.withUnknownLeader(logEndOffsetAndEpoch.epoch));
+        writeElectionStateToFile(file, stateFromMetadataLog);
+        return stateFromMetadataLog;
+    }
+
+    /**
+     * Reads the election state from cached quorum data.
+     */
+    @Override
+    public ElectionState readElectionState() {
+        return new ElectionState(
+            cachedState.leaderEpoch(),
+            cachedState.leaderId() == UNKNOWN_LEADER_ID ? OptionalInt.empty() :
+            OptionalInt.of(cachedState.leaderId()),
+            cachedState.votedId() == NOT_VOTED ? OptionalInt.empty() :
+            OptionalInt.of(cachedState.votedId()));
+    }
+
+    @Override
+    public void writeElectionState(ElectionState latest) throws IOException {
+        incorporateElectionState(cachedState, latest);
+        writeElectionStateToFile(stateFile, cachedState);
+    }
+
+    private void incorporateElectionState(QuorumStateData state, ElectionState latest) {
+        state.setLeaderEpoch(latest.epoch)
+            .setVotedId(latest.hasVoted() ? latest.votedId() : NOT_VOTED)
+            .setLeaderId(latest.hasLeader() ? latest.leaderId() : UNKNOWN_LEADER_ID);
+    }
+
+    private void writeElectionStateToFile(final File stateFile, QuorumStateData state) throws IOException  {
+        final File temp = new File(stateFile.getAbsolutePath() + ".tmp");
+        Files.deleteIfExists(temp.toPath());
+
+        log.trace("Writing tmp quorum state {}", temp.getAbsolutePath());
+        final FileOutputStream fileOutputStream = new FileOutputStream(temp);
+
+        try (final BufferedWriter writer = new BufferedWriter(
+            new OutputStreamWriter(fileOutputStream, StandardCharsets.UTF_8))) {
+            short version = state.highestSupportedVersion();
+
+            ObjectNode jsonState = (ObjectNode) state.toJson(version);
+            jsonState.set(DATA_VERSION, new ShortNode(version));
+            writer.write(jsonState.toString());
+            writer.flush();
+            fileOutputStream.getFD().sync();
+            Utils.atomicMoveWithFallback(temp.toPath(), stateFile.toPath());
+        } finally {
+            // cleanup the temp file when the write finishes (either success or fail).
+            Files.deleteIfExists(temp.toPath());
+        }
+    }
+
+    private String readLine(final BufferedReader reader) throws IOException {
```

Comment:
nit: do we really need a separate private function? I feel it is straight-forward just to make it inlined in the caller.